### PR TITLE
Allow zero length ranges in sanitization

### LINKED ISSUE #1607: [Regression] HarfBuzz rejects fvar table with 0 named instances after 587d49fc657c10
_(Thanks to http://underware.nl/ for reporting this bug.)_

https://github.com/harfbuzz/harfbuzz/commit/587d49fc657c10c8a20f2409a04d72bf80bb361e line 112 changed the sanitisation rules for the trail end of `fvar`. 

Further down in that change:
`get_first_instance ()` was changed to `get_instance (unsigned int i)`

in which  

```
  inline const InstanceRecord * get_first_instance (void) const	  inline const InstanceRecord &get_instance (unsigned int i) const
  { return &StructAfter<InstanceRecord> (get_axes ()); }
```

was changed to

```
  inline const InstanceRecord &get_instance (unsigned int i) const
  { return &StructAfter<InstanceRecord> (get_axes ()); }	  {
    if (unlikely (i >= instanceCount)) return Null (InstanceRecord);

    return StructAtOffset<InstanceRecord> (&StructAfter<InstanceRecord> (get_axes ()),
					  i * instanceSize);
  }
```

I am assuming that for a font with no named instances the latter does return a `Null (InstanceRecord)` which makes the 
` c->check_range (&get_instance (0), instanceCount, instanceSize));` sanitisation step fail.

If I remove `if (unlikely (i >= instanceCount)) return Null (InstanceRecord);` it starts accepting the test font's fvar again.

My interpretation of https://docs.microsoft.com/en-us/typography/opentype/spec/fvar is that no named instances are acceptable, so I consider this a regression which over-eagerly rejects fonts with no named instances.

### PR DESCRIPTION
**Pull request by ebraminio**
Fixes #1607

Closes #1616 also

And sorry for the white space noise. Let me add a test case also.

### CODE REVIEW

File: src/hb-machinery.hh [hunk 1]
```diff
@@ -326,37 +326,39 @@ struct hb_sanitize_context_t :
...
   bool check_range (const void *base,
-			   unsigned int len) const
+		    unsigned int len) const
   {
     const char *p = (const char *) base;
-    bool ok = this->start <= p &&
-	      p <= this->end &&
-	      (unsigned int) (this->end - p) >= len &&
-	      this->max_ops-- > 0;
+    bool ok = len == 0 ||
```
Comment (behdad) [inline src/hb-machinery.hh]: I personally prefer `!len` to `len == 0`, but am fine with this.

**Comment by ebraminio** [comment]
Now has a test also which didn't work before the test.

File: test/api/test-ot-face.c [hunk 2]
```diff
@@ -110,12 +110,21 @@ test_ot_face_empty (void)
   test_face (hb_face_get_empty (), 0);
 }
 
+static void
+test_ot_var_axis_on_zero_named_instance ()
+{
+  hb_face_t *face = hb_test_open_font_file ("fonts/Zycon.ttf");
```
Comment (ebraminio) [inline test/api/test-ot-face.c]: Zycon was available somewhere else on the repo and as duplicating file costs nothing on git I am copying and using it for the test.

**Comment by ebraminio** [comment]
Bots are happy, may I merge it?

**Comment by behdad** [comment]
Yep. Thanks!

**Comment by drott** [comment]
Thanks for the fix, great!

**Pull request merged**